src/compiler/utilities.ts

```diff
@@ -2801,13 +2810,7 @@ namespace ts {
                 return (<NamedDeclaration | PropertyAccessExpression>parent).name === node;
             case SyntaxKind.QualifiedName:
                 // Name on right hand side of dot in a type query or type reference
-                if ((<QualifiedName>parent).right === node) {
-                    while (parent.kind === SyntaxKind.QualifiedName) {
-                        parent = parent.parent;
-                    }
-                    return parent.kind === SyntaxKind.TypeQuery || parent.kind === SyntaxKind.TypeReference;
-                }
-                return false;
+                return (<QualifiedName>parent).right === node;

```

Comment:
I'm inferring that this change means

1. QualifiedNames are only ever children of TypeQuery/TypeReference
2. Unlike before, only top-level qualified names are identifier names, or else identifier names are parsed such that only top-level qualified names are actually passed to isIdentifierName.

Am I right?

Reply:
That is incorrect. The function is testing whether `node` is an _IdentifierName_ in ES. The _IdentifierName_ production is generally used for property names (i.e. `({ foo: 1 })` or `obj.foo`), where reserved words are not forbidden. A `QualifiedName` is a TS-only syntax, but is essentially similar to a property access expression when used as an expression. The name of a `QualifiedName` should always be considered an _IdentifierName_, even when it is not the child of a `TypeQuery` or `TypeReference`. The case this addresses is this:

```ts
export {};
import X = someNamespace.await;
```

Without this change, we would error on `await` since the namespace import is at the top level of a module, which is incorrect.

Reply:
Actually, *all* identifiers are _IdentifierName_ in ES. The difference is that there are essentially 4 categories of identifiers:

- [_IdentifierReference_](https://tc39.es/ecma262/#prod-IdentifierReference) - An identifier that is a _PrimaryExpression_ (e.g., `id` in `id()`, `id.prop`, `id + 1`, etc.).
- [_BindingIdentifier_](https://tc39.es/ecma262/#prod-BindingIdentifier) - An identifier that introduces a binding in the current scope (e.g., `id` in `var id`, `function id() {}`, `class id {}`, etc.).
- [_LabelIdentifier_](https://tc39.es/ecma262/#prod-LabelIdentifier) - An identifier that introduces a new label for use with `break`/`continue` (e.g., `id` in `id: while(true) break id`, etc.).
- Any **other** _IdentifierName_ such as the ones used in [_LiteralPropertyName_](https://tc39.es/ecma262/#prod-LiteralPropertyName) and [_MemberExpression_](https://tc39.es/ecma262/#prod-MemberExpression), [_ExportSpecifier_](https://tc39.es/ecma262/#prod-ExportSpecifier).

There is only one caller to this function, and what it is actually checking are these [static semantics](https://tc39.es/ecma262/#sec-identifiers-static-semantics-early-errors):

> _IdentifierReference_: `yield`\
> _BindingIdentifier_: `yield`\
> _LabelIdentifier_: `yield`
>
> It is a Syntax Error if the code matched by this production is contained in **strict mode** code.
>
>
> _IdentifierReference_: `await`\
> _BindingIdentifier_: `await`\
> _LabelIdentifier_: `await`
>
> It is a Syntax Error if the goal symbol of the syntactic grammar is _Module_.
>
> _BindingIdentifier_<sub>[Yield, Await]</sub>: `yield`
>
> It is a Syntax Error if this production has a `[Yield]` parameter.
>
> _BindingIdentifier_<sub>[Yield, Await]</sub>: `await`
>
> It is a Syntax Error if this production has an `[Await]` parameter.
>
> _IdentifierReference_<sub>[Yield, Await]</sub>: _Identifier_\
> _BindingIdentifier_<sub>[Yield, Await]</sub>: _Identifier_\
> _LabelIdentifier_<sub>[Yield, Await]</sub>: _Identifier_
>
> It is a Syntax Error if this production has a `[Yield]` parameter and StringValue of Identifier is `"yield"`.\
> It is a Syntax Error if this production has an `[Await]` parameter and StringValue of Identifier is `"await"`.

So what this function is actually testing is that the node is in the *other* category.